linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -2349,11 +2349,19 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
         val List(tarray, tindex) = targs
         tarray.tpe.base match {
           case arrayTpe @ ArrayType(ArrayTypeRef(base, _)) =>
-            val array = finishTransformExpr(tarray)
-            val index = finishTransformExpr(tindex)
             val elemType = cursoryArrayElemType(arrayTpe)
-            contTree(ArraySelect(array, index)(elemType))
-
+            if (!tarray.tpe.isNullable) {
+              val array = finishTransformExpr(tarray)
+              val index = finishTransformExpr(tindex)
+              val select = ArraySelect(array, index)(elemType)
+              contTree(select)
+            } else {
+              withNewTempLocalDefs(targs) { (localDefs, cont1) =>
+                val List(arrayDef, indexDef) = localDefs
+                val select = ArraySelect(arrayDef.newReplacement, indexDef.newReplacement)(elemType)
```

Comment:
The semantics of this node are very counter-intuitive, but specified as such:
https://lampwww.epfl.ch/~doeraene/sjsir-semantics/#sec-sjsir-arrayget-runtime-semantics-evaluation

I assume the reasons for it being specified that way lay somewhere at the intersection of how the relevant JVM and JS constructs are specified.

In any case, I think this warrants a comment, *why* the shorter form is valid if `tarray` is not-nullable.

Reply:
> I assume the reasons for it being specified that way lay somewhere at the intersection of how the relevant JVM and JS constructs are specified.

Yes. In JS, `array[index]` throws a `TypeError` before evaluating `index` when `array === null`. We have something similar for method calls.